openstack/blockstorage/v3/volumes/utils.go

```diff
@@ -45,3 +47,20 @@ func IDsFromName(client *gophercloud.ServiceClient, name string) ([]string, erro
 
 	return IDs, nil
 }
+
+// SafeVolumeDelete is a convenience function that returns an error if a volume can't be removed because it's in a wrong
+// state (e.g. it's attached to a server).
+func SafeVolumeDelete(client *gophercloud.ServiceClient, id string, opts volumes.DeleteOpts) error {
+	vol, err := volumes.Get(client, id).Extract()
+	if err != nil {
+		return err
+	}
+	// https://docs.openstack.org/api-ref/block-storage/v3/index.html?expanded=delete-a-volume-detail#delete-a-volume
+	accepted_states := []string{"available", "in-use", "error", "error_restoring", "error_extending", "error_managing"}
+	for _, state := range accepted_states {
```

Comment:
nit pick, but if you made the states a map instead of an array, then you could go from O(n) to O(1) :)

Reply:
Nice! I'd even say:

```Go
switch status := vol.Status {
case "available", "in-use", "error", "error_restoring", "error_extending", "error_managing":
	return volumes.Delete(client, id, opts).ExtractErr()
default:
	return fmt.Errorf("volume %s is in %s state, can't delete", id, status)
}
```